src/graph.rs

```diff
@@ -206,46 +207,45 @@ impl Graph {
     pub fn color_scc_par(&mut self) {
         log::trace!("Start coloring scc");
         let start = Instant::now();
-        let mut colors = vec![0; self.num_vertices];
-        let mut old_colors;
+        let mut colors = repeat_with(|| AtomicUsize::new(0))
+            .take(self.num_vertices)
+            .collect::<Vec<_>>();
 
         while !self.is_empty_par() {
-            colors
-                .par_iter_mut()
-                .enumerate()
-                .for_each(|(v, color)| match self.removed[v] {
-                    true => *color = self.scc_id[v],
-                    false => *color = v,
-                });
-
-            old_colors = colors.clone();
-
-            let color_changed = AtomicBool::new(true);
-            while color_changed.load(Ordering::Relaxed) {
-                color_changed.store(false, Ordering::Relaxed);
-
-                (0..self.num_vertices)
-                    .into_par_iter()
-                    .zip(colors.par_iter_mut())
-                    .zip(self.removed.par_iter())
-                    .filter_map(|((u, color), removed)| (!removed).then(|| (u, color)))
-                    .for_each(|(u, color)| {
-                        let range = self.csc.range_for(u);
-
-                        self.csc.unc[range]
-                            .iter()
-                            .filter(|w| !self.removed[**w])
-                            .for_each(|w| {
-                                if *color > old_colors[*w] {
```

Comment:
`colors` might have changed in other thread. Using `old_colors` will dramatically increase loop count.

Reply:
Thanks for the suggestion, removing the `old_colors` vector and producing the `color_changed` flag by reducing in parallel improves the performance a lot. 
It's not actually a logic error because the coloring loop will run again if the `color_changed` flag turned true by any thread.
See page 5 in the 'Coloring' Section of [this paper](https://www.cs.rpi.edu/~slotag/pub/SCC-IPDPS14.pdf)

> We place the parent in the queue to avoid explicit locking.
It is possible that two parents will have higher colors than
a shared child, creating a race condition. Both parents will
once again examine their children in the next iteration to
make sure that either the color that was given by them,
or a higher one, has been placed. Additionally, since only
a higher color can be assigned, we can ignore the race
condition created if a parent has their own color overwritten
before they assign their previous one to the child.

Reply:
Well, it won't produce wrong result, but it's not the same logic represented in the paper.